staging/src/k8s.io/apiextensions-apiserver/test/integration/fixtures/webhook.go

```diff
@@ -0,0 +1,198 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package fixtures
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/http/httptest"
+	"strings"
+	"sync"
+	"testing"
+	"time"
+
+	apiextensionsv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/uuid"
+)
+
+// WaitReadyFunc calls triggerConversionFn periodically and waits until it detects that the webhook
+// conversion server has handled at least 1 conversion request or the timeout is exceeded, in which
+// case an error is returned.
+type WaitReadyFunc func(timeout time.Duration, triggerConversionFn func() error) error
+
+// StartConversionWebhookServerWithWaitReady starts an http server with the provided handler and returns the WebhookClientConfig
+// needed to configure a CRD to use this conversion webhook as its converter.
+// It also returns a WaitReadyFunc to be called after the CRD is configured to wait until the conversion webhook handler
+// accepts at least one conversion request. If the server fails to start, an error is returned.
+// WaitReady is useful when changing the conversion webhook config of an existing CRD because the update does not take effect immediately.
+func StartConversionWebhookServerWithWaitReady(handler http.Handler) (func(), *apiextensionsv1beta1.WebhookClientConfig, WaitReadyFunc, error) {
+	var once sync.Once
+	handlerReadyC := make(chan struct{})
+	readyNotifyHandler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		once.Do(func() {
+			close(handlerReadyC)
+		})
+		handler.ServeHTTP(w, r)
+	})
+
+	tearDown, webhookConfig, err := StartConversionWebhookServer(readyNotifyHandler)
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("error starting webhook server: %v", err)
+	}
+
+	waitReady := func(timeout time.Duration, triggerConversionFn func() error) error {
+		var err error
+		for {
+			select {
+			case <-handlerReadyC:
+				return nil
+			case <-time.After(timeout):
+				return fmt.Errorf("Timed out waiting for CRD webhook converter update, last trigger conversion error: %v", err)
+			case <-time.After(100 * time.Millisecond):
+				err = triggerConversionFn()
+			}
+		}
+	}
+	return tearDown, webhookConfig, waitReady, err
+}
+
+// StartConversionWebhookServer starts an http server with the provided handler and returns the WebhookClientConfig
+// needed to configure a CRD to use this conversion webhook as its converter.
+func StartConversionWebhookServer(handler http.Handler) (func(), *apiextensionsv1beta1.WebhookClientConfig, error) {
+	// use unique paths for each webhook server to avoid accidentally having converters serve requests not intended for them
```

Comment:
old conversion hitting new path

The scenario I was thinking about here is when a CRD is updated to use a different webhook and we need our test to wait until the conversion client config has resulted in an updated rest storage before running the remainder of the test. I've written `StartConversionWebhookServerWithWaitReady` to make it possible to wait until a conversion webhook receives at least one request. This only waits long enough if the first request is is not from a old conversion client config. Not handling on the same path as the old conversion client config is a simple way to guarantee that.

Reply:
> we need our test to wait until the conversion client config has resulted in an updated rest storage before running the remainder of the test

this is about new conversion hitting old path / conversion config taking effect

> old conversion hitting new path

using a different path doesn't guarantee old conversion to finish before webhook server gets upgraded. It only makes the test fail hard if it raced. Also graceful storage deletion doesn't block webhook server upgrade

we guarantee we don't race between finishing old conversions and server upgrade because we run them sequentially. Maybe update the comment saying changing the path helps detect test failing instead of silently passing if the guarantee doesn't hold in future (e.g. old server teardown happens in parallel with CRUD tests)

Reply:
> > we need our test to wait until the conversion client config has resulted in an updated rest storage before running the remainder of the test
> 
> this is about new conversion hitting old path / conversion config taking effect

I don't think I understood what you meant by "old conversion" and "new conversion". But yes, this is about knowing when conversion config has taken effect so we can safely run tests that expect it.

> we guarantee we don't race between finishing old conversions and server upgrade because we run them sequentially. 

It's possible someone might write a test in the future that breaks this sequential assumption. But if they do I think we'd need to be able to better simulate starting/stopping conversion webhooks concurrently than we can with this utility.

> Maybe update the comment saying changing the path helps detect test failing instead of silently passing if the guarantee doesn't hold in future (e.g. old server teardown happens in parallel with CRUD tests)

Updated the comment to clarify a couple things.